File: Difficulty/Medium/[2] Add Two Numbers/solution.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#
# https://leetcode.com/problems/add-two-numbers/description/
#
# algorithms
# Medium (30.63%)
# Total Accepted:    821K
# Total Submissions: 2.7M
# Testcase Example:  '[2,4,3]\n[5,6,4]'
#
# You are given two non-empty linked lists representing two non-negative
# integers. The digits are stored in reverse order and each of their nodes
# contain a single digit. Add the two numbers and return it as a linked list.
# 
# You may assume the two numbers do not contain any leading zero, except the
# number 0 itself.
# 
# Example:
# 
# 
# Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
# Output: 7 -> 0 -> 8
# Explanation: 342 + 465 = 807.
# 
# 
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = l3 = ListNode(0)
        carry_flag = 0
        while( l1 and l2 ):
            l3.next = ListNode(0)
            l3 = l3.next
            temp = l1.val + l2.val + carry_flag
            if temp < 10:
                l3.val = temp
                carry_flag = 0
            else:
                l3.val = temp - 10
                carry_flag = 1
            l1 = l1.next
            l2 = l2.next
        if l1:
            l3.next = l1
            carry_flag, l3 = self.control_carry(carry_flag, l3)
        if l2:
            l3.next = l2
            carry_flag, l3 = self.control_carry(carry_flag, l3)
        if not l1 and not l2:
            if carry_flag == 1:
                l3.next = ListNode(1)
        if carry_flag == 1:
            l3.next = ListNode(1)
        return head.next

    def control_carry(self, carry_flag, l3):
        while(carry_flag):
            if l3.next:
                l3 = l3.next
            else:
                break
            temp1 = l3.val + carry_flag
            if temp1 < 10:
                l3.val = temp1
                carry_flag = 0
            else:
                l3.val = temp1 - 10
                carry_flag = 1
        return carry_flag, l3

File: Difficulty/Medium/[2] Add Two Numbers/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public int carry = 0;
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {        
        return addTwoNumbersHelper(l1, l2);
    }

    public ListNode addTwoNumbersHelper(ListNode l1, ListNode l2) {
        if (l1 == null) return processCarry(l2);
        if (l2 == null) return processCarry(l1);
        ListNode root = new ListNode(0);
        int val = l1.val + l2.val + carry;
        root.val = val % 10;
        carry = val / 10;
        root.next = addTwoNumbersHelper(l1.next, l2.next);
        return root;
    }

    public ListNode processCarry(ListNode l) {
        if (carry == 0) return l;
        if (l == null) {
            return new ListNode(1);
        } else {
            ListNode prev = l, tmp = l;
            while (tmp != null && carry == 1) {
                int val = tmp.val + carry;
                tmp.val = val % 10;
                carry = val / 10;
                prev = tmp;
                tmp = tmp.next;
            }
            if (carry == 1) {
                prev.next = new ListNode(1);
            }
        }
        return l;
    }
}